### PR TITLE
admission: Split stats for WorkQueueMetrics

### PR DESCRIPTION
**Pull request by andrewbaptist**
This commit splits the WorkQueueMetric stats into priority.
Because there are a lot of new stats that might be generated
as a result of this splitting, only the "important" priorities
are collected and considered for each queue.

Release note: None

### CODE REVIEW

**Review by sumeerbhola — CHANGES_REQUESTED** [review]
> Reviewed 5 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88076)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist and @irfansharif)

___
*[`pkg/server/server.go` line 222 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCq3vOU2EQc9rjooITN:-NCq3vOU2EQc9rjooITO:b4ongik) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd9d695320fa93f00889ac25dac28fafbc8b4b6/pkg/server/server.go#L222)):*
> ```Go
> 		admissionOptions.Override(opts)
> 	}
> 	admissionOptions.Settings = st
> ```

why this change to pull out settings from the options?
___
*[`pkg/ts/catalog/chart_catalog.go` line 3473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCq40e-6-ic0afH2dXz:-NCq40e-6-ic0afH2dY-:b-emgb6q) ([raw file](https://github.com/cockroachdb/cockroach/blob/905c8848c211889301a60f24d9a30e46332ea9ab/pkg/ts/catalog/chart_catalog.go#L3473)):*
> ```Go
> 					"admission.errored.kv.0",
> 					"admission.errored.kv.100",
> 					"admission.errored.sql-kv-response.0",
> ```

is this 0 and 100 the priority values?
I am wondering why we've picked these to populate here and not the others in the `admissionpb.WorkPriority` enum.
___
*[`pkg/util/admission/grant_coordinator.go` line 158 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCq0Skk7y6o1iIjfzhv:-NCq0Skk7y6o1iIjfzhw:b-kfsehl) ([raw file](https://github.com/cockroachdb/cockroach/blob/905c8848c211889301a60f24d9a30e46332ea9ab/pkg/util/admission/grant_coordinator.go#L158)):*
> ```Go
> 	// This is IO work, so override the usesTokens value.
> 	opts.usesTokens = true
> 	// TODO(sumeer): add per-store WorkQueue state for debug.zip and db console.
> ```

this revised code is also sharing the WorkQueue metrics across all stores, yes?
This would be a good thing to fix in a followup PR.
___
*[`pkg/util/admission/grant_coordinator.go` line 970 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCq13D235_mLr60TWsS:-NCq13D235_mLr60TWsT:b-vuqirc) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd9d695320fa93f00889ac25dac28fafbc8b4b6/pkg/util/admission/grant_coordinator.go#L970)):*
> ```Go
> 
> // Prevent the linter from emitting unused warnings.
> var _ = NewGrantCoordinatorSQL
> ```

why is this removed?
___
*[`pkg/util/admission/work_queue.go` line 1598 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCq35c2EkFUjF63Q__Z:-NCq35c2EkFUjF63Q___:bqqkea7) ([raw file](https://github.com/cockroachdb/cockroach/blob/905c8848c211889301a60f24d9a30e46332ea9ab/pkg/util/admission/work_queue.go#L1598)):*
> ```Go
> 	// TODO(abaptist): This is done to pre-register stats. Need to check that we
> 	// getOrCreate "enough" of the priorities to be useful.
> 	for _, pri := range applicablePriorities {
> ```

if the lazy creation is working correctly, this eager creation seems unnecessary. Is is needed for the list in chart_catalog?
___
*[`pkg/kv/kvserver/client_split_test.go` line 1224 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCq1aNEC7rekbJUTRp3:-NCq1aNEC7rekbJUTRp4:bqebhii) ([raw file](https://github.com/cockroachdb/cockroach/blob/905c8848c211889301a60f24d9a30e46332ea9ab/pkg/kv/kvserver/client_split_test.go#L1224)):*
> ```Go
> 			// Set maxBytes to something small so we can exceed the maximum split
> 			// size without adding 2x64MB of data.
> 			const maxBytes = 1 << 17
> ```

why this change?

**Review by andrewbaptist — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88076)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @sumeerbhola)

___
*[`pkg/server/server.go` line 222 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCq3vOU2EQc9rjooITN:-NCu1dNZ4h7x52f_9P3z:b-9rt76d) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd9d695320fa93f00889ac25dac28fafbc8b4b6/pkg/server/server.go#L222)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

why this change to pull out settings from the options?
</blockquote></details>

There was inconsistency in the NewGrantCoordinator in terms of how it processed the settings. This wasn't strictly necessary, however, it made it harder to follow the code in terms of where the settings vs options came from. The goal was really to have more immutable options and settings coming through this code to make reasoning and testing easier. I don't mind reverting this part of the patch (or pulling into another commit). 
___
*[`pkg/ts/catalog/chart_catalog.go` line 3473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCq40e-6-ic0afH2dXz:-NCu2Bex936Er2Fwrl5R:beaquc8) ([raw file](https://github.com/cockroachdb/cockroach/blob/905c8848c211889301a60f24d9a30e46332ea9ab/pkg/ts/catalog/chart_catalog.go#L3473)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

is this 0 and 100 the priority values?
I am wondering why we've picked these to populate here and not the others in the `admissionpb.WorkPriority` enum.
</blockquote></details>

These were the values that were most commonly used (0, 100 and -30). Once lazy loading is in place in the future, ideally this won't be necessary to do.
___
*[`pkg/util/admission/grant_coordinator.go` line 158 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCq0Skk7y6o1iIjfzhv:-NCu2PUrEQS1CcezpKMQ:b5e808f) ([raw file](https://github.com/cockroachdb/cockroach/blob/905c8848c211889301a60f24d9a30e46332ea9ab/pkg/util/admission/grant_coordinator.go#L158)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

this revised code is also sharing the WorkQueue metrics across all stores, yes?
This would be a good thing to fix in a followup PR.
</blockquote></details>

This comment seemed out of place after this chance the metrics were more obviously inserted as part of the `makeStoreRequesterFunc`. 
___
*[`pkg/util/admission/grant_coordinator.go` line 970 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCq13D235_mLr60TWsS:-NCu2iEuAQaor63_dD4J:b-2y0lot) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd9d695320fa93f00889ac25dac28fafbc8b4b6/pkg/util/admission/grant_coordinator.go#L970)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

why is this removed?
</blockquote></details>

Not removed, just moved up to right before the method to make it more obvious what it was referring to. The `NewGrantCoordinatorSQL` method will need a little work before it is enabled, and it took me a while to understand how it worked (since it was repeating metrics names) as I didn't see that it was unused. 
___
*[`pkg/util/admission/work_queue.go` line 1598 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCq35c2EkFUjF63Q__Z:-NCu3EDwA2zZnGgeDcmw:bu0x9c) ([raw file](https://github.com/cockroachdb/cockroach/blob/905c8848c211889301a60f24d9a30e46332ea9ab/pkg/util/admission/work_queue.go#L1598)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

if the lazy creation is working correctly, this eager creation seems unnecessary. Is is needed for the list in chart_catalog?
</blockquote></details>

lazy creation is working correctly from the metrics perspective and they get registered, unfortunately the registry does not currently handle lazy addtion (see https://cockroachlabs.slack.com/archives/C01CNRP6TSN/p1663779962687259 )
___
*[`pkg/kv/kvserver/client_split_test.go` line 1224 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCq1aNEC7rekbJUTRp3:-NCu0o8QCoibSpCeO1gg:b5hsn3h) ([raw file](https://github.com/cockroachdb/cockroach/blob/905c8848c211889301a60f24d9a30e46332ea9ab/pkg/kv/kvserver/client_split_test.go#L1224)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

why this change?
</blockquote></details>

The two tests started failing after making the change to stats. This test initializes a system with a greatly reduced `maxBytes` and then attempts to create a `maxBytes` size value and check what happens on split. With the new stats, we push the metrics value over the 64KB limit and therefore this test fails because it also gets added to the split queue. The 1 <<16 was somewhat arbitrary, so this change simply makes it large enough so the metrics range is no longer included.

**Review by sumeerbhola — APPROVED** [review]
:lgtm: 

> Reviewed 4 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-:-NCuDjO0AzOix_PlFK91:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andrewbaptist, @irfansharif, and @sumeerbhola)

___
*[`-- commits` line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCuDdq2-tnkiy6MXmdA):*
Please add something like
Informs #82743
___
*[`pkg/server/server.go` line 222 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCq3vOU2EQc9rjooITN:-NCu8VJACMjVxdHdtszY:bb9e6st) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd9d695320fa93f00889ac25dac28fafbc8b4b6/pkg/server/server.go#L222)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

There was inconsistency in the NewGrantCoordinator in terms of how it processed the settings. This wasn't strictly necessary, however, it made it harder to follow the code in terms of where the settings vs options came from. The goal was really to have more immutable options and settings coming through this code to make reasoning and testing easier. I don't mind reverting this part of the patch (or pulling into another commit). 
</blockquote></details>

btw, I don't think we have a principle regarding separating immutable and mutable configuration -- we tend to pile them all into Config/Options structs. Maybe kv code is principled in this regard, in which case I'd also like to know what the idiom is. One example is `base.StorageConfig` which includes `cluster.Settings`.

But I am ok with your change.
___
*[`pkg/ts/catalog/chart_catalog.go` line 3473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCq40e-6-ic0afH2dXz:-NCuAORYDsqXOecWk_tY:b-tg4c2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/905c8848c211889301a60f24d9a30e46332ea9ab/pkg/ts/catalog/chart_catalog.go#L3473)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

These were the values that were most commonly used (0, 100 and -30). Once lazy loading is in place in the future, ideally this won't be necessary to do.
</blockquote></details>

So will we not see metrics for -30, -50, -100 for kv, kv-stores etc. given that the lazy registration does not work?
If yes, we should eagerly register them. One of the reasons we wanted to change is to use it when diagnosing workloads where we have user-facing stuff (typically priority 0) and various bulk operations running concurrently (and the latter will use -30 and -100). 
___
*[`pkg/util/admission/grant_coordinator.go` line 158 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCq0Skk7y6o1iIjfzhv:-NCuBxV46dV9UJnrHtRw:b-q038mh) ([raw file](https://github.com/cockroachdb/cockroach/blob/905c8848c211889301a60f24d9a30e46332ea9ab/pkg/util/admission/grant_coordinator.go#L158)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

This comment seemed out of place after this chance the metrics were more obviously inserted as part of the `makeStoreRequesterFunc`. 
</blockquote></details>

nit: `makeStoreRequesterFunc` is accepting metrics created in `makeStoresGrantCoordinators` and not creating them itself, so we need a TODO comment about per-store metrics somewhere in `StoreGrantCoordinators`. See my comment below about a better place.
___
*[`pkg/util/admission/grant_coordinator.go` line 401 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCuCUs9CVPiJAkSdNlm:-NCuCUs9CVPiJAkSdNln:b-mdn8g0) ([raw file](https://github.com/cockroachdb/cockroach/blob/905c8848c211889301a60f24d9a30e46332ea9ab/pkg/util/admission/grant_coordinator.go#L401)):*
> ```Go
> 	// These metrics are shared across all stores and all
> 	// priorities across stores (even the coarser workClasses, which are a
> 	// mapping from priority, share the same metrics).
> ```

Comment needs updating since the "all priorities" part is stale now. Can you add a TODO here about per-store metrics?
___
*[`pkg/util/admission/grant_coordinator.go` line 970 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCq13D235_mLr60TWsS:-NCuCzCeCJVQ-mfy3y8H:b2nhb2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd9d695320fa93f00889ac25dac28fafbc8b4b6/pkg/util/admission/grant_coordinator.go#L970)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Not removed, just moved up to right before the method to make it more obvious what it was referring to. The `NewGrantCoordinatorSQL` method will need a little work before it is enabled, and it took me a while to understand how it worked (since it was repeating metrics names) as I didn't see that it was unused. 
</blockquote></details>

Ack

File: pkg/util/admission/granter_test.go
Comment (irfansharif) [inline pkg/util/admission/granter_test.go:124]: Typo.

File: pkg/util/admission/work_queue.go
Comment (irfansharif) [inline pkg/util/admission/work_queue.go:1526]: Incomplete sentence.

File: pkg/util/admission/work_queue.go
Comment (irfansharif) [inline pkg/util/admission/work_queue.go:1537]: With this lazy registration -- does these metrics show up in CRDB's internal timeseries database? I'm asking whether it works with everything under pkg/server/status/recorder.go since a lot of that code was written with upfront registration in mind.

File: pkg/util/admission/work_queue.go
Comment (irfansharif) [inline pkg/util/admission/work_queue.go:1598]: So it seems we're doing pre-registration here anyway. Do we need the LoadOrStore above when finding the right metric? If everything is properly pre-registered, shouldn't we fatal if we don't find something in the sync.Map?

**Review by andrewbaptist — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88076)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @irfansharif and @sumeerbhola)

___
*[`-- commits` line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCuDdq2-tnkiy6MXmdA):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Please add something like
Informs #82743
</blockquote></details>

Done
___
*[`pkg/ts/catalog/chart_catalog.go` line 3468 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCLg6soOlJNLIkyz4Q9-r1-3468:-NCvFKCeB9TEVevT5fdh:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/905c8848c211889301a60f24d9a30e46332ea9ab/pkg/ts/catalog/chart_catalog.go#L3468)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Instead of the numerical suffix, can we use the name of priority level instead? I can see us needing to slot in other priorities numerically without wanting that to reflect in different metric names/updating UI/warranting a release note. It's also easier to understand what these specific metrics represent without needing to look at code.
</blockquote></details>

Done
___
*[`pkg/ts/catalog/chart_catalog.go` line 3473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCq40e-6-ic0afH2dXz:-NCvFQKI84UmYUCGSNQ5:bjwzxo7) ([raw file](https://github.com/cockroachdb/cockroach/blob/905c8848c211889301a60f24d9a30e46332ea9ab/pkg/ts/catalog/chart_catalog.go#L3473)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

So will we not see metrics for -30, -50, -100 for kv, kv-stores etc. given that the lazy registration does not work?
If yes, we should eagerly register them. One of the reasons we wanted to change is to use it when diagnosing workloads where we have user-facing stuff (typically priority 0) and various bulk operations running concurrently (and the latter will use -30 and -100). 
</blockquote></details>

Done, I added some additional stats, I am attempting to avoid overloading the number of stats. Having a few missed stats isn't too bad because they will fall into the "default" bucket. 
___
*[`pkg/util/admission/grant_coordinator.go` line 158 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCq0Skk7y6o1iIjfzhv:-NCvFIKt9Y_cbUfMXF5u:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/905c8848c211889301a60f24d9a30e46332ea9ab/pkg/util/admission/grant_coordinator.go#L158)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: `makeStoreRequesterFunc` is accepting metrics created in `makeStoresGrantCoordinators` and not creating them itself, so we need a TODO comment about per-store metrics somewhere in `StoreGrantCoordinators`. See my comment below about a better place.
</blockquote></details>

Done
___
*[`pkg/util/admission/grant_coordinator.go` line 401 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCuCUs9CVPiJAkSdNlm:-NCv9BtuEchWI96s3sCh:b-896fjx) ([raw file](https://github.com/cockroachdb/cockroach/blob/905c8848c211889301a60f24d9a30e46332ea9ab/pkg/util/admission/grant_coordinator.go#L401)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Comment needs updating since the "all priorities" part is stale now. Can you add a TODO here about per-store metrics?
</blockquote></details>

Done
___
*[`pkg/util/admission/grant_coordinator.go` line 970 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCq13D235_mLr60TWsS:-NCv9FyEBfRHu78iVUoM:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd9d695320fa93f00889ac25dac28fafbc8b4b6/pkg/util/admission/grant_coordinator.go#L970)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Ack
</blockquote></details>

Done
___
*[`pkg/util/admission/work_queue.go` line 1526 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NC7Ksn0bVAa7iLbluD5-r1-1526:-NCv9RI657UpJ4K9SoI3:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/905c8848c211889301a60f24d9a30e46332ea9ab/pkg/util/admission/work_queue.go#L1526)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Incomplete sentence.
</blockquote></details>

Done
___
*[`pkg/util/admission/work_queue.go` line 1537 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NC7Ksn0bVAa7iLbluD5-r1-1537:-NCv9Z2L6tG5Hl9rlusD:b-27rpeb) ([raw file](https://github.com/cockroachdb/cockroach/blob/905c8848c211889301a60f24d9a30e46332ea9ab/pkg/util/admission/work_queue.go#L1537)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

With this lazy registration -- does these metrics show up in CRDB's internal timeseries database? I'm asking whether it works with everything under pkg/server/status/recorder.go since a lot of that code was written with upfront registration in mind.
</blockquote></details>

No - the lazy registration today does not handle either the internal or prometheus database. I hope we can remove the pre-registration in the future, but I didn't dive into that code now. 
___
*[`pkg/util/admission/work_queue.go` line 1598 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NC7Ksn0bVAa7iLbluD5-r1-1598:-NCv9yNz06zL-hGWS0qd:bcvwrfd) ([raw file](https://github.com/cockroachdb/cockroach/blob/905c8848c211889301a60f24d9a30e46332ea9ab/pkg/util/admission/work_queue.go#L1598)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

So it seems we're doing pre-registration here anyway. Do we need the LoadOrStore above when finding the right metric? If everything is properly pre-registered, shouldn't we fatal if we don't find something in the sync.Map?
</blockquote></details>

We don't currently pre-register everything so this shouldn't be fatal. I want to remove pre-registration in the future, but didn't want to try and do this as part of this commit. I kept LoadOrStore for that reason.
___
*[`pkg/util/admission/granter_test.go` line 124 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NC7Ksn0bVAa7iLbluD4-r1-124:-NCvAQwn31JQcIatsGfA:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/905c8848c211889301a60f24d9a30e46332ea9ab/pkg/util/admission/granter_test.go#L124)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Typo.
</blockquote></details>

Done

**Review by sumeerbhola — COMMENTED** [review]
> Reviewed 4 of 5 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88076)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andrewbaptist, @irfansharif, and @sumeerbhola)

___
*[`pkg/util/admission/work_queue.go` line 1529 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCzZNH66t0RqJWOGBjr:-NCzZNH66t0RqJWOGBjs:bkz6qfn) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd9d695320fa93f00889ac25dac28fafbc8b4b6/pkg/util/admission/work_queue.go#L1529)):*
> ```Go
> 	Admitted        *metric.Counter
> 	Errored         *metric.Counter
> 	WaitDurationSum *metric.Counter
> ```

You are removing this because we don't find it useful, yes?
My original motivation was to be able to look at mean latency, but it is not possible to compute delta(waitDurationSum)/delta(requested) in our internal metrics system, though it it possible with Prometheus.
I'll leave it to @irfansharif and your judgement on whether this is useful or not.

**Review by andrewbaptist — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88076)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @irfansharif and @sumeerbhola)

___
*[`pkg/util/admission/work_queue.go` line 1529 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCzZNH66t0RqJWOGBjr:-NCzb6xC-Ij0BzW_Adzx:bj6d5no) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd9d695320fa93f00889ac25dac28fafbc8b4b6/pkg/util/admission/work_queue.go#L1529)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

You are removing this because we don't find it useful, yes?
My original motivation was to be able to look at mean latency, but it is not possible to compute delta(waitDurationSum)/delta(requested) in our internal metrics system, though it it possible with Prometheus.
I'll leave it to @irfansharif and your judgement on whether this is useful or not.

</blockquote></details>

This is automatically computed as part of the WaitDurations histogram, and the values are the same. So this was a redundant state. I'm not sure about the internal metrics, but I assume they could use the sum metric. Prometheus can definitely use the histogram counter to get the sum

**Review by irfansharif — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88076)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andrewbaptist, @irfansharif, and @sumeerbhola)

___
*[`pkg/ts/catalog/chart_catalog.go` line 3463 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-ND-UObZ8aNRbgNm4Ij4:-ND-UObZ8aNRbgNm4Ij5:b-dum7g2) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d3e20916d6fcad3a734f81c09cdae28e23456c6/pkg/ts/catalog/chart_catalog.go#L3463)):*
> ```Go
> 					"admission.admitted.elastic-cpu",
> 					"admission.errored.elastic-cpu",
> 					"admission.admitted.elastic-cpu.bulk",
> ```

Optional nit: add the "-pri" suffix so it's even clearer.
___
*[`pkg/util/admission/work_queue.go` line 1529 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCzZNH66t0RqJWOGBjr:-ND-UEjk89mgl_0DxY6-:b-azl6u9) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd9d695320fa93f00889ac25dac28fafbc8b4b6/pkg/util/admission/work_queue.go#L1529)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

This is automatically computed as part of the WaitDurations histogram, and the values are the same. So this was a redundant state. I'm not sure about the internal metrics, but I assume they could use the sum metric. Prometheus can definitely use the histogram counter to get the sum
</blockquote></details>

https://github.com/cockroachdb/cockroach/blob/e397faf3918d5757fd22b21a6a0ee76e72074ea0/pkg/server/status/recorder.go#L532-L541

The sum is missing, but the avg is computed already. So this is redundant and can be removed.

**Comment by andrewbaptist** [comment]
bors r=irfansharif

**Comment by andreimatei** [comment]
This patch increased the number of timeseries exported to tsdb by crdb nodes from 731 to 1739. If there's histograms in there, it might be even much worse for Prometheus.
I don't think we can afford anything like it...

**Comment by andrewbaptist** [comment]
I was concerned about the impact of this change. I can easily disable all of these and will hold off on any backporting. I might need to rethink if there is a way to get the data we want without exploding stats.

**Comment by andreimatei** [comment]
I think I fooled myself when counting, sorry. I now think that this patch only took the timeseries count from 1543 to 1739. So, now quite as bad as I thought. Still significant, though.

**Review by sumeerbhola — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88076)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[`pkg/util/admission/work_queue.go` line 1529 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88076#-NCzZNH66t0RqJWOGBjr:-NDY_P7UC4dk7VFQj7ES:bcxw1fg) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd9d695320fa93f00889ac25dac28fafbc8b4b6/pkg/util/admission/work_queue.go#L1529)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

https://github.com/cockroachdb/cockroach/blob/e397faf3918d5757fd22b21a6a0ee76e72074ea0/pkg/server/status/recorder.go#L532-L541

The sum is missing, but the avg is computed already. So this is redundant and can be removed.
</blockquote></details>

The histogram reporting is limited to work that actually waited in the queue. So if a large fraction did not wait in the queue, it will overreport the latency percentiles. We should fix that by reporting everything. I did not do this when writing the first version of this code, because I was worried about lock contention overhead in the histogram and did not have time to investigate how much of a problem it really was.

**Pull request merged**
